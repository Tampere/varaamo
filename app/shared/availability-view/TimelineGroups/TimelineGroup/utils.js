import { slotSize, slotWidth, slotMargin } from 'constants/SlotConstants';

import some from 'lodash/some';
import moment from 'moment';

function getTimeSlotWidth({ startTime, endTime } = {}) {
  const diff = endTime ? endTime.diff(startTime, 'minutes') : slotSize;
  const slots = Math.floor(diff / slotSize);

  return (slotWidth * slots) - slotMargin;
}

function getTimelineItems(date, reservations, resourceId) {
  const items = [];
  let reservationPointer = 0;
  let timePointer = date.clone().startOf('day');
  const end = date.clone().endOf('day');
  while (timePointer.isBefore(end)) {
    const reservation = reservations && reservations[reservationPointer];
    const isSlotReservation = reservation && timePointer.isSame(reservation.begin);
    if (isSlotReservation) {
      items.push({
        key: String(items.length),
        type: 'reservation',
        data: reservation,
      });
      timePointer = moment(reservation.end);
      reservationPointer += 1;
    } else {
      items.push({
        key: String(items.length),
        type: 'reservation-slot',
        data: {
          begin: timePointer.format(),
          end: timePointer.clone().add(slotSize, 'minutes').format(),
          resourceId,
          // isSelectable: false by default to improve selector performance by allowing
          // addSelectionData to make some assumptions.
          isSelectable: false,
        },
      });
      timePointer.add(slotSize, 'minutes');
    }
  }
  return items;
}

function isInsideOpeningHours(item, openingHours) {
  return some(openingHours, opening => (
    opening.opens <= item.data.begin && item.data.end <= opening.closes
  ));
}

function markItemSelectable(item, isSelectable, openingHours) {
  const selectable = (
    isSelectable
    && moment().isSameOrBefore(item.data.end)
    && (!openingHours || isInsideOpeningHours(item, openingHours))
  );
  return { ...item, data: { ...item.data, isSelectable: selectable } };
}

function markItemsSelectable(items, isSelectable, openingHours) {
  return items.map((item) => {
    if (item.type === 'reservation') return item;
    return markItemSelectable(item, isSelectable, openingHours);
  });
}

function addSelectionData(selection, resource, items) {
  if (!selection) {
    return markItemsSelectable(items, true, resource.openingHours);
  } if (selection.resourceId !== resource.id) {
    // isSelectable is false by default, so nothing needs to be done.
    // This is a pretty important performance optimization when there are tons of
    // resources in the AvailabilityView and the selection is in a state where the
    // first click has been done but the second (end time) hasn't. Without this
    // optimization we'd be calling markItemSelectable for every slot in every
    // resource when the user hovers to another slot.
    return items;
  }
  let lastSelectableFound = false;
  return items.map((item) => {
    if (lastSelectableFound || item.data.begin < selection.begin) {
      if (item.type === 'reservation') return item;
      // isSelectable is false by default.
      return item;
    }
    if (item.type === 'reservation') {
      lastSelectableFound = true;
      return item;
    }
    return markItemSelectable(item, true, resource.openingHours);
  });
}

export default {
  addSelectionData,
  getTimelineItems,
  getTimeSlotWidth,
};
