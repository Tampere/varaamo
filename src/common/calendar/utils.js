import moment from 'moment';
import get from 'lodash/get';

/**
 * getDefaultSelectedTimeRange()
 * Get default selected time range from reservation data.
 * @returns {Object} selected time range
 */
export const getDefaultSelectedTimeRange = (reservation) => {
  return reservation ? {
    start: reservation.begin,
    end: reservation.end,
  } : null;
};


/**
 * isTimeRangeOverMaxPeriod();
 * Reservation cannot be longer than the resources max period if max period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than max_period or not
 */
export const isTimeRangeOverMaxPeriod = (resource, start, end, isStaff = false) => {
  const period = get(resource, 'max_period', null);

  if (!isStaff && period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') > durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * isTimeRangeUnderMinPeriod();
 * Reservation must longer than the resources min period if min period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than min_period or not
 */
export const isTimeRangeUnderMinPeriod = (resource, start, end) => {
  const period = get(resource, 'min_period', null);

  if (period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') < durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * addPeriodToTime()
 * Add period to input time
 * For example: period 00:30:00, time: moment(12:00:00) -> moment(12:30:00)
 *
 * @param {Moment} time
 * @param {string} period
 * @returns {Moment}
 */

const addPeriodToTime = (time, period) => {
  if (period) {
    const periodInMinute = moment.duration(period).as('minutes');
    return time.add(periodInMinute, 'minutes');
  }

  // Return original if no period
  return time;
};

/**
 * getMinPeriodTimeRange
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMinPeriodTimeRange = (resource, start, end) => {
  const minPeriod = get(resource, 'min_period', null);

  if (!minPeriod) {
    return {
      start, end,
    };
  }

  const minPeriodEnd = addPeriodToTime(moment(start), minPeriod);

  if (minPeriodEnd.diff(moment(end)) > 0) {
    // time range smaller than min_period, return min_period
    return {
      start,
      end: minPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};

/**
 * getMaxPeriodEndTime
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMaxPeriodTimeRange = (resource, start, end, isStaff) => {
  const maxPeriod = get(resource, 'max_period', null);

  if (!maxPeriod || isStaff) {
    return {
      start, end,
    };
  }

  const maxPeriodEnd = addPeriodToTime(moment(start), maxPeriod);

  if (moment(end).diff(maxPeriodEnd) > 0) {
    // time range is over max_period not allowed, return max_period
    return {
      start,
      end: maxPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};
