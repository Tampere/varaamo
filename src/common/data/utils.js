import get from 'lodash/get';
import values from 'lodash/values';
import pickBy from 'lodash/pickBy';
import camelCase from 'lodash/camelCase';

export const getLocalizedFieldValue = (field, locale, fallback = false) => {
  const localeValue = get(field, locale, null);

  if (localeValue || !fallback) {
    return localeValue;
  }

  return values(field).find(fallbackValue => !!fallbackValue);
};

/**
 * Trim empty values from data object
 *
 * @param {Object} data
 * @returns {Object} Trimmed empty values object
 */

export const parseData = (data) => {
  return pickBy(data, value => value || value === 0);
};

export const toCamelCase = (obj) => {
  if (!obj) return {};

  const camelCasedObj = {};

  Object.keys(obj).forEach((key) => {
    const camelCasedKey = camelCase(key);
    camelCasedObj[camelCasedKey] = obj[key];
  });

  return camelCasedObj;
};
