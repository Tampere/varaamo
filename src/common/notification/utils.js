import { NotificationManager } from 'react-notifications';

import { NOTIFICATION_TYPE } from './constants';
/**
 * createNotification()
 * Creating notification popups
 *
 * @param {string} type
 * @param {string} [message='']
 * @param {string} [title='']
 * @param {number} [fadeTime=2000]
 * @param {function} callback
 * @returns {void}
 */
export const createNotification = (type, message = '', title = '', fadeTime = 2000, callback) => {
  switch (type) {
    case NOTIFICATION_TYPE.INFO:
      NotificationManager.info(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.SUCCESS:
      NotificationManager.success(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.WARNING:
      NotificationManager.warning(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.ERROR:
      NotificationManager.error(message, title, fadeTime, callback);
      break;
    default:
      NotificationManager.info(message, title, fadeTime, callback);
  }
};
