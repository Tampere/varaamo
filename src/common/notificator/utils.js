import moment from 'moment';

/**
 * createSelectOptions()
 * Create option arrays for selects.
 * @returns {Object} with select option arrays
 */
export const createSelectOptions = () => {
  const targetOptions = [
    { value: 'staff', label: 'Staff' },
    { value: 'user', label: 'User' },
    { value: 'all', label: 'All' },
  ];
  const urgencyOptions = [
    { value: 'common', label: 'Common' },
    { value: 'warning', label: 'Warning' },
    { value: 'danger', label: 'Danger' },
  ];
  const activeOptions = [
    { value: true, label: 'True' },
    { value: false, label: 'False' },
  ];

  return {
    targetOptions,
    urgencyOptions,
    activeOptions,
  };
};


/**
 * formatValuesForUse()
 * Format notification values to usable form.
 * @returns {Object} with formatted values
 */
export const formatValuesForUse = (notification) => {
  const { targetOptions, urgencyOptions, activeOptions } = createSelectOptions();
  const formattedNotification = { ...notification };
  formattedNotification.target = targetOptions.find(option => option.value === notification.target);
  formattedNotification.urgency = urgencyOptions.find(option => option.value === notification.urgency);
  formattedNotification.active = activeOptions.find(option => option.value === notification.active);
  formattedNotification.until = moment(notification.until).toDate();
  return formattedNotification;
};

/**
 * formatValuesForDatabase()
 * Format notification values to savable form.
 * @returns {Object} with formatted values
 */
export const formatValuesForDatabase = (notification) => {
  const formattedNotification = { ...notification };
  formattedNotification.target = notification.target.value;
  formattedNotification.urgency = notification.urgency.value;
  formattedNotification.active = notification.active.value;
  formattedNotification.until = moment(notification.until).format('YYYYMMDDTHHmmss');
  return formattedNotification;
};
