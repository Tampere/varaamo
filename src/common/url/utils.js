/**
 * Builds search string from an object.
 * @param search {object}
 * @returns {string}
 */
export const getSearch = (search) => {
  return new URLSearchParams(search).toString();
};

/**
 * Getter for link string with given path and query params.
 * @param path {string}
 * @param search {object}
 * @returns {string}
 */
export const getLinkString = (path, search = null) => {
  let searchString = '';

  if (search) {
    searchString = getSearch(search);
  }

  return `${path}${search ? `?${searchString}` : ''}`;
};
