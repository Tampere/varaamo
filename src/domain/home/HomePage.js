import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Button, Col, Row } from 'react-bootstrap';
import Loader from 'react-loader';
import camelCase from 'lodash/camelCase';
import Link from 'react-router-dom/Link';

import HomeSearchBox from './HomeSearchBox';
import iconManufacturing from './images/frontpage_build.svg';
import iconPhotoAndAudio from './images/frontpage_music.svg';
import iconSports from './images/frontpage_sport.svg';
import iconGuidance from './images/frontpage_guidance.svg';
import iconMeetingsAndWorking from './images/frontpage_work.svg';
import iconOrganizeEvents from './images/frontpage_event.svg';
import iconSauna from './images/frontpage_sauna.svg';
import client from '../../common/api/client';
import FAIcon from '../../../app/shared/fontawesome-icon/FontAwesomeIcon';
import PageWrapper from '../../../app/pages/PageWrapper';
import injectT from '../../../app/i18n/injectT';

const purposeIcons = {
  photoAndAudio: iconPhotoAndAudio,
  sports: iconSports,
  guidance: iconGuidance,
  manufacturing: iconManufacturing,
  meetingsAndWorking: iconMeetingsAndWorking,
  events: iconOrganizeEvents,
  sauna: iconSauna,
};

class HomePage extends Component {
  state = {
    isFetchingPurposes: false,
    purposes: null,
  }

  componentDidMount() {
    this.fetchPurposes();
  }

  fetchPurposes = async () => {
    this.setState({ isFetchingPurposes: true });
    try {
      const response = await client.get('purpose');
      this.setState({ isFetchingPurposes: false, purposes: response.data.results });
    } catch (errors) {
      this.setState({ isFetchingPurposes: false });
      // TODO: HTTP error handler
    }
  }

  handleSearch = (value = '') => {
    this.props.history.push(`/search?search=${value}`);
  }

  renderPurposeBanners = () => {
    const { t, locale } = this.props;
    const { purposes } = this.state;

    const normalizedPurpose = purposes.filter(purpose => !purpose.parent).sort();

    return normalizedPurpose.map((purpose) => {
      const image = purposeIcons[camelCase(purpose.id)];
      const purposeName = purpose.name[locale] || '';

      return (
        <Col className="app-HomePageContent__banner" key={purpose.id} md={3} sm={6} xs={12}>
          <Link className="app-HomePageContent__banner__linkWrapper" to={`/search?purpose=${purpose.id}`}>
            <div className="app-HomePageContent__banner-icon">
              {typeof image === 'string' ? <img alt={purposeName} src={image} />
              // TODO: VAR-80 | VAR-81 Replace those icon with designed icon.

                : <FAIcon icon={image} />}
            </div>

            <h5>{purposeName}</h5>
            <div className="app-HomePageContent__banner-action">
              <Button
                bsStyle="primary"
                className="app-HomePageContent__button"
              >
                {t('HomePage.buttonText')}
              </Button>
            </div>
          </Link>
        </Col>
      );
    });
  }

  render() {
    const { t } = this.props;
    const { purposes, isFetchingPurposes } = this.state;

    return (
      <div className="app-HomePage">
        <div className="app-HomePage__content container">
          <h1>Varaamo â€“</h1>
          <h1>{t('HomePage.contentTitle')}</h1>
          <h5>{t('HomePage.contentSubTitle')}</h5>
          <HomeSearchBox onSearch={this.handleSearch} />
        </div>
        <div className="app-HomePage__koro" />
        <PageWrapper className="app-HomePageContent" title={t('HomePage.title')}>
          <h4>{t('HomePage.bannersTitle')}</h4>
          {purposes && (
            <Loader loaded={!isFetchingPurposes}>
              <div className="app-HomePageContent__banners">
                <Row>
                  {this.renderPurposeBanners()}
                </Row>
              </div>
            </Loader>
          )}
        </PageWrapper>
      </div>
    );
  }
}

HomePage.propTypes = {
  history: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
  locale: PropTypes.string.isRequired,
};

export default injectT(HomePage);
