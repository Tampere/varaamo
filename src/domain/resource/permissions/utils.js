import difference from 'lodash/difference';
import get from 'lodash/get';

import {
  resourceRoles, resourcePermissionTypes, resourcePermissionsByRole, UI_UNIT_STAFF_ROLES,
} from './constants';

const permissionOptions = Object.values(resourcePermissionTypes);
const allValuesInArray = (subset, superset) => difference(subset, superset).length === 0;

export function hasPermissionForResource(resourceRole, requiredPermissions) {
  // undefined roles don't receive any permissions implicitly
  if (!resourceRole) {
    return false;
  }

  if (!Object.values(resourceRoles).includes(resourceRole)) {
    throw Error(`resourceRole needs to be one of: ${Object.values(resourceRoles).join(', ')}.`);
  }

  if (!Array.isArray(requiredPermissions) && typeof requiredPermissions !== 'string') {
    throw Error('requiredPermissions has to be either a string or an array.');
  }

  if (
    Array.isArray(requiredPermissions)
    && !allValuesInArray(requiredPermissions, permissionOptions)
  ) {
    const unknownPermissions = difference(requiredPermissions, permissionOptions);

    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${unknownPermissions.join(', ')}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  if (typeof requiredPermissions === 'string' && !permissionOptions.includes(requiredPermissions)) {
    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${requiredPermissions}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  const rolePermissions = resourcePermissionsByRole[resourceRole];

  if (Array.isArray(requiredPermissions)) {
    return allValuesInArray(requiredPermissions, rolePermissions);
  }

  return rolePermissions.includes(requiredPermissions);
}

const getUserPermissions = resource => (
  get(resource, 'user_permissions', false)
  || get(resource, 'userPermissions', undefined)
);

// Returns the role with the most permissions or null.
export function getUnitRoleFromResource(resource) {
  const userPermissions = getUserPermissions(resource);

  if (!userPermissions) {
    return null;
  }

  if (userPermissions.is_admin || userPermissions.isAdmin) {
    return resourceRoles.UNIT_ADMINISTRATOR;
  }

  if (userPermissions.is_manager || userPermissions.isManager) {
    return resourceRoles.UNIT_MANAGER;
  }

  if (userPermissions.is_viewer || userPermissions.isViewer) {
    return resourceRoles.UNIT_VIEWER;
  }

  return null;
}

export function getIsUnitStaff(unitRole) {
  return UI_UNIT_STAFF_ROLES.includes(unitRole);
}
