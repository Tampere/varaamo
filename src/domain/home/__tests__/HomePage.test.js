import React from 'react';
import simple from 'simple-mock';
import Link from 'react-router-dom/Link';
import toJson from 'enzyme-to-json';

import HomePage from '../HomePage';
import { shallowWithIntl } from '../../../../app/utils/testUtils';

describe('pages/home/HomePage', () => {
  const history = {
    push: () => {},
  };

  const defaultProps = {
    history,
    isFetchingPurposes: false,
    locale: 'en',
  };

  function getWrapper(extraProps) {
    return shallowWithIntl(<HomePage {...defaultProps} {...extraProps} />);
  }

  describe('render', () => {
    describe('Purpose banners', () => {
      let wrapper;

      beforeAll(() => {
        wrapper = getWrapper();
      });

      afterAll(() => {
        simple.restore();
      });

      test('render normally', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
      });

      test(' have at least a Link component', () => {
        wrapper.setState({
          purposes: [
            {
              id: 'purpose-foo',
              name: {
                en: 'purpose-name',
              },
            },
          ],
        });

        wrapper.update();
        expect(wrapper.find(Link).first().prop('to')).toContain('purpose-foo');
      });
    });
  });

  describe('handleSearch', () => {
    const value = 'some value';
    const expectedPath = `/search?search=${value}`;
    let historyMock;

    beforeAll(() => {
      const instance = getWrapper().instance();
      historyMock = simple.mock(history, 'push');
      instance.handleSearch(value);
    });

    afterAll(() => {
      simple.restore();
    });

    test('calls browserHistory push with correct path', () => {
      expect(historyMock.callCount).toBe(1);
      expect(historyMock.lastCall.args).toEqual([expectedPath]);
    });
  });
});
