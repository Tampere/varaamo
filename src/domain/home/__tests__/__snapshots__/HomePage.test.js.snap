// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/home/HomePage render Purpose banners render normally 1`] = `
<div
  className="app-HomePage"
>
  <div
    className="app-HomePage__content container"
  >
    <h1>
      Varaamo â€“
    </h1>
    <h1>
      HomePage.contentTitle
    </h1>
    <h5>
      HomePage.contentSubTitle
    </h5>
    <InjectT(HomeSearchBox)
      onSearch={[Function]}
    />
  </div>
  <div
    className="app-HomePage__koro"
  />
  <PageWrapper
    className="app-HomePageContent"
    title="HomePage.title"
  >
    <h4>
      HomePage.bannersTitle
    </h4>
  </PageWrapper>
</div>
`;
